### PR TITLE
Added kubelet config source from apiserver.

### PR DESCRIPTION
**Pull request by erictune**

### CODE REVIEW

**Comment by erictune** [comment]
part of #2483

**Comment by erictune** [comment]
@pmorie

**Comment by erictune** [comment]
@thockin you were working with sources too.

**Comment by erictune** [comment]
@saad-ali you were working in kubelet sources.  This is adding one.  PTAL to see if I need to pickup your changes in #3355 in my added file.

**Comment by saad-ali** [comment]
Thanks for the heads up @erictune. I'll sync up with you tomorrow.

**Comment by erictune** [comment]
I added a test to ensure that when List returns an empty list, an update is still sent to kubelet.  @saad-ali said this is what is necessary to handle bug from #3355.

File: pkg/kubelet/config/apiserver.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/watch"
+	"github.com/golang/glog"
+)
+
+type podLW interface {
```
Comment (lavalamp) [inline pkg/kubelet/config/apiserver.go]: Is this interface needed? Can you just use cache.ListerWatcher?

**Comment by erictune** [comment]
This is ready for oncall to assign a reviewer.

File: pkg/kubelet/config/apiserver.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	}
+	newSourceApiserverFromLW(lw, updates)
+}
+
+// newSourceApiserverFromLW holds creates a config source that watches an pulls from the apiserver.
+func newSourceApiserverFromLW(lw podLW, updates chan<- interface{}) {
+	send := func(objs []interface{}) {
+		var bpods []api.BoundPod
+		for _, o := range objs {
+			pod := o.(*api.Pod)
```
Comment (lavalamp) [inline pkg/kubelet/config/apiserver.go]: I think you need to use the bound pod factory from pkg/registry/pod...

**Comment by lavalamp** [comment]
If calling the bound pod factory proves difficult, you may need to simultaneously with this change wire real pods through the kubelet...

File: pkg/kubelet/config/apiserver.go (hunk 1 above)
Comment (erictune) [inline pkg/kubelet/config/apiserver.go]: done.

File: pkg/kubelet/config/apiserver.go (hunk 2 above)
Comment (erictune) [inline pkg/kubelet/config/apiserver.go]: BoundPodFactory can't be easily reused.  And I plan to delete it once this series of changes finishes.

I reviewed what MakeBoundPod does:
  - `api.Scheme.Convert` -- doing that already
  - set env vars -- I intend this to be done in a different part of kubelet in a different PR
  - sets selflink -- Need to add this.

Comment (erictune) [inline pkg/kubelet/config/apiserver.go]: added self link

**Comment by lavalamp** [comment]
LGTM, but travis is broken

**Comment by erictune** [comment]
PTAL

**Comment by lavalamp** [comment]
LGTM

**Pull request merged**